File: sdks/go/cmd/elydora/plugins/codex_contract.go
package plugins

import (
	"bytes"
	"encoding/json"
	"fmt"
	"path/filepath"
	"reflect"
	"sort"
)

const (
	codexAgentKey           = "codex"
	codexConfigFile         = "hooks.json"
	codexGuardScript        = "guard.js"
	codexAuditScript        = "hook.js"
	codexHookTimeout        = float64(10)
	codexOwnedDescription   = "Elydora audit and freeze enforcement"
	codexGuardStatusMessage = "Checking Elydora agent state"
	codexAuditStatusMessage = "Recording Elydora tool use"
)

type codexHooks map[string][]map[string]any

type codexDocument struct {
	exists   bool
	filePath string
	root     map[string]any
	hooks    codexHooks
	raw      []byte
}

type codexRenderedDocument struct {
	document *codexDocument
	changed  bool
	next     []byte
	remove   bool
}

type codexRuntimeContract = managedRuntimeContract

func cloneCodexHooks(source codexHooks) codexHooks {
	clone := make(codexHooks, len(source))
	for event, groups := range source {
		clone[event] = append([]map[string]any(nil), groups...)
	}
	return clone
}

func readCodexHooks(value any, label string) (codexHooks, error) {
	object, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf(`%s field "hooks" must be an object`, label)
	}
	hooks := make(codexHooks, len(object))
	for event, candidate := range object {
		values, ok := candidate.([]any)
		if !ok {
			return nil, fmt.Errorf(`%s field "hooks.%s" must be an array`, label, event)
		}
		groups := make([]map[string]any, 0, len(values))
		for groupIndex, value := range values {
			group, ok := value.(map[string]any)
			if !ok || group == nil {
				return nil, fmt.Errorf(
					`%s matcher group hooks.%s[%d] must be an object`,
					label, event, groupIndex,
				)
			}
			handlerValues, ok := group["hooks"].([]any)
			if !ok {
				return nil, fmt.Errorf(
					`%s matcher group hooks.%s[%d] must contain a hooks array`,
					label, event, groupIndex,
				)
			}
			for handlerIndex, handler := range handlerValues {
				if object, ok := handler.(map[string]any); !ok || object == nil {
					return nil, fmt.Errorf(
						`%s handler hooks.%s[%d].hooks[%d] must be an object`,
						label, event, groupIndex, handlerIndex,
					)
				}
			}
			groups = append(groups, group)
		}
		hooks[event] = groups
	}
	return hooks, nil
}

func parseCodexDocument(filePath string, raw []byte) (*codexDocument, error) {
	label := fmt.Sprintf("Codex user hooks at %s", filePath)
	root, err := decodeStrictJSONObject(raw, label)
	if err != nil {
		return nil, err
	}
	hooks := codexHooks{}
	if value, exists := root["hooks"]; exists {
		hooks, err = readCodexHooks(value, label)
		if err != nil {
			return nil, err
		}
	}
	return &codexDocument{
		exists: true, filePath: filePath, root: root,
		hooks: hooks, raw: append([]byte(nil), raw...),
	}, nil
}

func createCodexDocument(filePath string) *codexDocument {
	return &codexDocument{
		filePath: filePath,
		root:     map[string]any{"description": codexOwnedDescription},
		hooks:    codexHooks{},
	}
}

func exactCodexMatcherGroup(group map[string]any) bool {
	return len(group) == 2 && group["matcher"] == "*"
}

func removeCodexGroups(
	groups []map[string]any,
	scriptName string,
	status string,
	agentID string,
) []map[string]any {
	result := make([]map[string]any, 0, len(groups))
	for _, group := range groups {
		handlers := group["hooks"].([]any)
		kept := make([]any, 0, len(handlers))
		for _, value := range handlers {
			handler := value.(map[string]any)
			managedID, managed := codexManagedAgentID(handler, scriptName, status)
			if managed && (agentID == "" || sameManagedAgentID(managedID, agentID)) {
				continue
			}
			kept = append(kept, handler)
		}
		if len(kept) > 0 || !exactCodexMatcherGroup(group) {
			next := cloneJSONObject(group)
			next["hooks"] = kept
			result = append(result, next)
		}
	}
	return result
}

func removeManagedCodexHooks(hooks codexHooks, agentID string) codexHooks {
	next := cloneCodexHooks(hooks)
	for _, contract := range []struct{ event, script, status string }{
		{"PreToolUse", codexGuardScript, codexGuardStatusMessage},
		{"PostToolUse", codexAuditScript, codexAuditStatusMessage},
	} {
		groups := removeCodexGroups(
			next[contract.event], contract.script, contract.status, agentID,
		)
		if len(groups) == 0 {
			delete(next, contract.event)
		} else {
			next[contract.event] = groups
		}
	}
	return next
}

func entirelyManagedCodexDocument(document *codexDocument) bool {
	if !document.exists || document.root["description"] != codexOwnedDescription ||
		len(document.hooks) == 0 {
		return false
	}
	for key := range document.root {
		if key != "description" && key != "hooks" {
			return false
		}
	}
	handlerCount := 0
	for event, groups := range document.hooks {
		contract := map[string][2]string{
			"PreToolUse":  {codexGuardScript, codexGuardStatusMessage},
			"PostToolUse": {codexAuditScript, codexAuditStatusMessage},
		}[event]
		if contract[0] == "" || len(groups) == 0 {
			return false
		}
		for _, group := range groups {
			handlers := group["hooks"].([]any)
			if !exactCodexMatcherGroup(group) || len(handlers) == 0 {
				return false
			}
			for _, value := range handlers {
				if _, managed := codexManagedAgentID(
					value.(map[string]any), contract[0], contract[1],
				); !managed {
					return false
				}
				handlerCount++
			}
		}
	}
	return handlerCount > 0
}

func renderCodexDocument(
	document *codexDocument,
	hooks codexHooks,
) (*codexRenderedDocument, error) {
	if !document.exists && len(hooks) == 0 {
		return &codexRenderedDocument{document: document}, nil
	}
	if document.exists && reflect.DeepEqual(hooks, document.hooks) {
		return &codexRenderedDocument{document: document}, nil
	}
	if len(hooks) == 0 && entirelyManagedCodexDocument(document) {
		return &codexRenderedDocument{document: document, changed: true, remove: true}, nil
	}
	root := cloneJSONObject(document.root)
	if len(hooks) == 0 {
		delete(root, "hooks")
	} else {
		root["hooks"] = hooks
	}
	next, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return nil, err
	}
	next = append(next, '\n')
	return &codexRenderedDocument{
		document: document,
		changed:  !document.exists || !bytes.Equal(next, document.raw),
		next:     next,
	}, nil
}

func managedCodexIDs(
	groups []map[string]any,
	scriptName string,
	status string,
) map[string]string {
	result := map[string]string{}
	for _, group := range groups {
		if !exactCodexMatcherGroup(group) {
			continue
		}
		for _, value := range group["hooks"].([]any) {
			agentID, managed := codexManagedAgentID(value.(map[string]any), scriptName, status)
			if !managed {
				continue
			}
			result[managedReferenceKey(agentID)] = agentID
		}
	}
	return result
}

func codexRuntimeContracts(hooks codexHooks) ([]codexRuntimeContract, error) {
	guards := managedCodexIDs(hooks["PreToolUse"], codexGuardScript, codexGuardStatusMessage)
	audits := managedCodexIDs(hooks["PostToolUse"], codexAuditScript, codexAuditStatusMessage)
	keys := make([]string, 0, len(guards))
	for key := range guards {
		if _, exists := audits[key]; exists {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)
	root, err := AgentRuntimeRoot()
	if err != nil {
		return nil, err
	}
	contracts := make([]codexRuntimeContract, 0, len(keys))
	for _, key := range keys {
		agentID := guards[key]
		directory := filepath.Join(root, agentID)
		contracts = append(contracts, codexRuntimeContract{
			agentID:   agentID,
			guardPath: filepath.Join(directory, codexGuardScript),
			auditPath: filepath.Join(directory, codexAuditScript),
		})
	}
	return contracts, nil
}

File: sdks/node/src/plugins/codex-contract.ts
import os from 'node:os';
import path from 'node:path';
import { sameAgentId, samePath } from './common.js';
import {
  encodedWindowsCommand,
  isNodeExecutable,
  parseEncodedWindowsCommand,
  parseLegacyWindowsCommand,
  parsePosixCommand,
  posixSource,
} from './shell-command.js';
import { isObject, parseStrictJsonObject, type JsonObject } from './strict-json.js';

export const AGENT_KEY = 'codex';
export const CONFIG_FILE = 'hooks.json';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_SECONDS = 10;
export const OWNED_DESCRIPTION = 'Elydora audit and freeze enforcement';
export const GUARD_STATUS = 'Checking Elydora agent state';
export const AUDIT_STATUS = 'Recording Elydora tool use';

export type CodexHooks = Record<string, JsonObject[]>;

export interface CodexDocument {
  readonly exists: boolean;
  readonly filePath: string;
  readonly root: JsonObject;
  readonly hooks: CodexHooks;
  readonly raw?: string;
}

export interface RenderedDocument {
  readonly document: CodexDocument;
  readonly changed: boolean;
  readonly next?: string;
}

export interface RuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

export function buildHandler(scriptPath: string, statusMessage: string): JsonObject {
  return {
    type: 'command',
    command: posixSource(scriptPath),
    commandWindows: encodedWindowsCommand(scriptPath),
    timeout: HOOK_TIMEOUT_SECONDS,
    statusMessage,
  };
}

function exactHandlerKeys(handler: JsonObject): boolean {
  return Object.keys(handler).sort().join('|')
    === 'command|commandWindows|statusMessage|timeout|type';
}

function managedScriptPath(handler: JsonObject, statusMessage: string): string | undefined {
  if (!exactHandlerKeys(handler)
    || handler.type !== 'command'
    || handler.timeout !== HOOK_TIMEOUT_SECONDS
    || handler.statusMessage !== statusMessage) return undefined;
  if (typeof handler.command !== 'string' || typeof handler.commandWindows !== 'string') {
    return undefined;
  }
  const posix = parsePosixCommand(handler.command);
  const windows = parseEncodedWindowsCommand(handler.commandWindows)
    ?? parseLegacyWindowsCommand(handler.commandWindows);
  if (!posix || !windows
    || !path.isAbsolute(posix[0])
    || !path.isAbsolute(posix[1])
    || !isNodeExecutable(posix[0])
    || !isNodeExecutable(windows[0])
    || !samePath(posix[0], windows[0])
    || !samePath(posix[1], windows[1])) return undefined;
  return posix[1];
}

function managedAgentId(
  handler: JsonObject,
  scriptName: string,
  statusMessage: string,
): string | undefined {
  const scriptPath = managedScriptPath(handler, statusMessage);
  if (!scriptPath || path.basename(scriptPath) !== scriptName) return undefined;
  const agentDirectory = path.dirname(scriptPath);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) {
    return undefined;
  }
  const agentId = path.basename(agentDirectory);
  return agentId && agentId !== '.' && agentId !== '..' ? agentId : undefined;
}

function readHooks(value: unknown, label: string): CodexHooks {
  if (value === undefined) return {};
  if (!isObject(value)) throw new Error(`${label} field "hooks" must be an object`);
  const hooks: CodexHooks = {};
  for (const [event, candidate] of Object.entries(value)) {
    if (!Array.isArray(candidate)) {
      throw new Error(`${label} field "hooks.${event}" must be an array`);
    }
    hooks[event] = candidate.map((group, groupIndex) => {
      if (!isObject(group)) {
        throw new Error(`${label} matcher group hooks.${event}[${groupIndex}] must be an object`);
      }
      if (!Array.isArray(group.hooks)) {
        throw new Error(
          `${label} matcher group hooks.${event}[${groupIndex}] must contain a hooks array`,
        );
      }
      for (const [handlerIndex, handler] of group.hooks.entries()) {
        if (!isObject(handler)) {
          throw new Error(
            `${label} handler hooks.${event}[${groupIndex}].hooks[${handlerIndex}] must be an object`,
          );
        }
      }
      return group;
    });
  }
  return hooks;
}

export function parseDocument(filePath: string, raw: string): CodexDocument {
  const label = `Codex user hooks at ${filePath}`;
  const root = parseStrictJsonObject(raw, label);
  return {
    exists: true,
    filePath,
    root,
    hooks: readHooks(root.hooks, label),
    raw,
  };
}

export function createDocument(filePath: string): CodexDocument {
  return {
    exists: false,
    filePath,
    root: { description: OWNED_DESCRIPTION },
    hooks: {},
  };
}

function removeFromGroups(
  groups: JsonObject[],
  scriptName: string,
  statusMessage: string,
  agentId?: string,
): JsonObject[] {
  const result: JsonObject[] = [];
  for (const group of groups) {
    const handlers = (group.hooks as JsonObject[]).filter((handler) => {
      const managedId = managedAgentId(handler, scriptName, statusMessage);
      return !managedId || (agentId !== undefined && !sameAgentId(managedId, agentId));
    });
    if (handlers.length > 0) result.push({ ...group, hooks: handlers });
    else if (!exactMatcherGroup(group)) result.push({ ...group, hooks: [] });
  }
  return result;
}

export function removeManagedHooks(hooks: CodexHooks, agentId?: string): CodexHooks {
  const next: CodexHooks = Object.fromEntries(
    Object.entries(hooks).map(([event, groups]) => [event, [...groups]]),
  );
  for (const [event, scriptName, statusMessage] of [
    ['PreToolUse', GUARD_SCRIPT, GUARD_STATUS],
    ['PostToolUse', AUDIT_SCRIPT, AUDIT_STATUS],
  ] as const) {
    const groups = removeFromGroups(next[event] ?? [], scriptName, statusMessage, agentId);
    if (groups.length > 0) next[event] = groups;
    else delete next[event];
  }
  return next;
}

function entirelyManaged(document: CodexDocument): boolean {
  if (!document.exists
    || document.root.description !== OWNED_DESCRIPTION
    || !Object.keys(document.root).every((key) => key === 'description' || key === 'hooks')) {
    return false;
  }
  const events = Object.entries(document.hooks);
  if (events.length === 0) return false;
  let handlerCount = 0;
  for (const [event, groups] of events) {
    const contract = event === 'PreToolUse'
      ? [GUARD_SCRIPT, GUARD_STATUS] as const
      : event === 'PostToolUse'
        ? [AUDIT_SCRIPT, AUDIT_STATUS] as const
        : undefined;
    if (!contract || groups.length === 0) return false;
    for (const group of groups) {
      const handlers = group.hooks as JsonObject[];
      if (!exactMatcherGroup(group)
        || handlers.length === 0
        || handlers.some((handler) => !managedAgentId(handler, contract[0], contract[1]))) {
        return false;
      }
      handlerCount += handlers.length;
    }
  }
  return handlerCount > 0;
}

export function renderDocument(
  document: CodexDocument,
  hooks: CodexHooks,
): RenderedDocument {
  if (!document.exists && Object.keys(hooks).length === 0) {
    return { document, changed: false };
  }
  if (Object.keys(hooks).length === 0 && entirelyManaged(document)) {
    return { document, changed: true };
  }
  const root: JsonObject = { ...document.root };
  if (Object.keys(hooks).length > 0) root.hooks = hooks;
  else delete root.hooks;
  const next = `${JSON.stringify(root, null, 2)}\n`;
  return { document, changed: next !== document.raw, next };
}

function exactMatcherGroup(group: JsonObject): boolean {
  return Object.keys(group).sort().join('|') === 'hooks|matcher' && group.matcher === '*';
}

function managedIds(
  groups: JsonObject[],
  scriptName: string,
  statusMessage: string,
): Map<string, string> {
  const result = new Map<string, string>();
  for (const group of groups) {
    if (!exactMatcherGroup(group)) continue;
    for (const handler of group.hooks as JsonObject[]) {
      const agentId = managedAgentId(handler, scriptName, statusMessage);
      if (!agentId) continue;
      const key = process.platform === 'win32' ? agentId.toLowerCase() : agentId;
      result.set(key, agentId);
    }
  }
  return result;
}

export function runtimeContracts(hooks: CodexHooks): RuntimeContract[] {
  const guards = managedIds(hooks.PreToolUse ?? [], GUARD_SCRIPT, GUARD_STATUS);
  const audits = managedIds(hooks.PostToolUse ?? [], AUDIT_SCRIPT, AUDIT_STATUS);
  const root = path.join(os.homedir(), '.elydora');
  return [...guards]
    .filter(([key]) => audits.has(key))
    .map(([, agentId]) => ({
      agentId,
      guardPath: path.join(root, agentId, GUARD_SCRIPT),
      auditPath: path.join(root, agentId, AUDIT_SCRIPT),
    }));
}
